scene/gui/line_edit.cpp

```diff
@@ -2109,10 +2109,26 @@ void LineEdit::delete_text(int p_from_column, int p_to_column) {
 	}
 }
 
+void LineEdit::_set_text_and_emit_text_changed_signal(String p_text) {
+	String previous_text = get_text();
+	set_text(p_text);
+
+	if (get_text() == previous_text) {
+		return;
+	}
+
+	_text_changed();
+}
+
 void LineEdit::set_text(String p_text) {
 	clear_internal();
+
+	String previous_text = get_text();
 	insert_text_at_caret(p_text);
-	_create_undo_state();
+
+	if (get_text() != previous_text) {
+		_create_undo_state();
+	}
```

Comment:
I left a "fix" here. Technically, we should always create an undo state only if the action was needed, isn't it?